### PR TITLE
AV-199351 update gateway status with rest layer error

### PR DESCRIPTION
**Pull request by arihantg**

### CODE REVIEW

File: ako-gateway-api/status/gateway_status.go [hunk 1]
```diff
@@ -140,31 +140,41 @@ func (o *gateway) Update(key string, option status.StatusOptions) {
...
 	addressType := gatewayv1.IPAddressType
-	status.Addresses = append(status.Addresses, gatewayv1.GatewayStatusAddress{
-		Type:  &addressType,
-		Value: option.Options.Vip[0],
-	})
-
-	// TODO: Add a way to propagate the error from the Rest layer to status layer.
-
+	if len(option.Options.Vip) != 0 {
+		status.Addresses = append(status.Addresses, gatewayv1.GatewayStatusAddress{
```
Comment (pkoshtavmware) [inline ako-gateway-api/status/gateway_status.go:144]: This append function will add a duplicate vip everytime we will send status update to the gateway

File: ako-gateway-api/status/gateway_status.go [hunk 2]
```diff
@@ -140,31 +140,41 @@ func (o *gateway) Update(key string, option status.StatusOptions) {
...
+			SetIn(&status.Conditions)
+
+		for i := range status.Listeners {
+			listenerCondition := NewCondition()
+			listenerCondition.
+				Type(string(gatewayv1.GatewayConditionProgrammed)).
+				Status(metav1.ConditionTrue).
+				Reason(string(gatewayv1.GatewayReasonProgrammed)).
+				ObservedGeneration(gw.ObjectMeta.Generation).
+				Message("Virtual service configured/updated").
```
Comment (pkoshtavmware) [inline ako-gateway-api/status/gateway_status.go:175]: I think same message should be appended here as well else gateway condition will show gateway is in error and listener condition will show virtual service configured or updated

File: ako-gateway-api/status/gateway_status.go [hunk 3]
```diff
@@ -140,31 +140,41 @@ func (o *gateway) Update(key string, option status.StatusOptions) {
...
-		ObservedGeneration(gw.ObjectMeta.Generation).
-		Message("Virtual service configured/updated").
-		SetIn(&status.Conditions)
 
-	for i := range status.Listeners {
-		listenerCondition := NewCondition()
-		listenerCondition.
+	if option.Options.Message != "" {
+		condition.
+			Type(string(gatewayv1.GatewayConditionProgrammed)).
```
Comment (pkoshtavmware) [inline ako-gateway-api/status/gateway_status.go:153]: Gateway Condition should be set to Accepted with gateway status false

**Review by pkoshtavmware — COMMENTED** [review]
I think changes might need to be done in delete case as well

File: ako-gateway-api/status/gateway_status.go (hunk 2 above)
Comment (arihantg) [reply ako-gateway-api/status/gateway_status.go]: appended the message in listener section

File: ako-gateway-api/status/gateway_status.go (hunk 3 above)
Comment (arihantg) [reply ako-gateway-api/status/gateway_status.go]: Changed gateway condition to accepted

**Comment by arihantg** [comment]
> I think changes might need to be done in delete case as well

delete case would also go through same flow once rest api gives error. No change is needed for that

**Comment by arihantg** [comment]
build ako

**Comment by arihantg** [comment]
build ako

**Comment by pkoshtavmware** [comment]
Overall PR looks good to me.
Could you please attach FT run ?

**Comment by arihantg** [comment]
Screenshot 2024-03-18 at 4 48 54 PM
Screenshot 2024-03-17 at 7 10 35 PM

**Review by akshayhavile — COMMENTED** [review]
@arihantg Overall changes looks good from a situation where Error is there while creating gateway. But if we look at the `Gateway yaml` provided in the issue, it should be accepted as these are valid configuration. We should fix AKO-gw sending duplicate value for port combination as part of this issue,

**Comment by pkoshtavmware** [comment]
> @arihantg Overall changes looks good from a situation where Error is there while creating gateway. But if we look at the `Gateway yaml` provided in the issue, it should be accepted as these are valid configuration. We should fix AKO-gw sending duplicate value for port combination as part of this issue,

Hi @akshayhavile ,
Isn't @aaha97 's PR taking care of this fix ?
Could you please confirm @aaha97 ?

**Review by pkoshtavmware — APPROVED** [review]
LGTM

**Review by akshayhavile — APPROVED** [review]

**Pull request merged**